staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go

```diff
@@ -169,48 +180,244 @@ func validateCertificateAuthority(certificateAuthority string, fldPath *field.Pa
+		celMapper.ClaimValidationRules = authenticationcel.NewMapper(compilationResults)
+	}
+
+	return allErrs
+}
+
+func validateClaimMappings(compiler authenticationcel.Compiler, celMapper *authenticationcel.CELMapper, m api.ClaimMappings, fldPath *field.Path, structuredAuthnFeatureEnabled bool) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if !structuredAuthnFeatureEnabled {
+		if m.Username.Expression != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("username").Child("expression"), *m.Username.Expression, "expression is not supported when StructuredAuthenticationConfiguration feature gate is disabled"))
+		}
+		if m.Groups.Expression != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("groups").Child("expression"), *m.Groups.Expression, "expression is not supported when StructuredAuthenticationConfiguration feature gate is disabled"))
+		}
+		if len(m.UID.Claim) > 0 || len(m.UID.Expression) > 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("uid"), "", "uid claim mapping is not supported when StructuredAuthenticationConfiguration feature gate is disabled"))
+		}
+		if len(m.Extra) > 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("extra"), "", "extra claim mapping is not supported when StructuredAuthenticationConfiguration feature gate is disabled"))
+		}
+		return allErrs
+	}
+
+	compilationResults, err := validatePrefixClaimOrExpression(compiler, m.Username, fldPath.Child("username"), true, structuredAuthnFeatureEnabled)
+	if err != nil {
+		allErrs = append(allErrs, err...)
+	} else if len(compilationResults) > 0 {
+		celMapper.Username = authenticationcel.NewMapper(compilationResults)
+	}
+
+	compilationResults, err = validatePrefixClaimOrExpression(compiler, m.Groups, fldPath.Child("groups"), false, structuredAuthnFeatureEnabled)
+	if err != nil {
+		allErrs = append(allErrs, err...)
+	} else if len(compilationResults) > 0 {
+		celMapper.Groups = authenticationcel.NewMapper(compilationResults)
+	}
+
+	switch {
+	case len(m.UID.Claim) > 0 && len(m.UID.Expression) > 0:
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("uid"), "", "claim and expression can't both be set"))
+	case len(m.UID.Expression) > 0:
+		compilationResult, err := validateCELCondition(compiler, &authenticationcel.ClaimMappingCondition{
+			Expression: m.UID.Expression,
+		}, authenticationcel.ClaimsVarName, fldPath.Child("uid").Child("expression"))
+
+		if err != nil {
+			allErrs = append(allErrs, err)
+		} else {
+			celMapper.UID = authenticationcel.NewMapper([]authenticationcel.CompilationResult{compilationResult})
+		}
+	}
+
+	var extraCompilationResults []authenticationcel.CompilationResult
+	// There is no check for duplicate extra mapping keys. If multiple
+	// mappings have the same key, the result will be a concatenation of
+	// all the values with the order preserved.
+	for i, mapping := range m.Extra {
+		fldPath := fldPath.Child("extra").Index(i)
+		if len(mapping.Key) == 0 {
```

Comment:
Validate that key is lowercase and namespaced per:

https://github.com/kubernetes/kubernetes/blob/3825e206cb162a7ad7431a5bdf6a065ae8422cf7/staging/src/k8s.io/apiserver/pkg/authentication/user/user.go#L31-L41